### PR TITLE
samples: Apply default configuration

### PR DESCRIPTION
**Pull request by rakons**
Add default configuration to the bluetooth and nfc examples.
All the examples has:
- Enable assertion
- Enable logger
- Enable stack protection

Signed-off-by: Radoslaw Koppel <radoslaw.koppel@nordicsemi.no>

### CODE REVIEW

**Comment by rakons** [comment]
The idea behind this commit is to align the output from the example. To limit the surprise to the user that one example prints something (log enabled) and the other can silently go to dead loop (assertion enabled, log disabled) by default.

Please look into yours examples and see if you may align the configuration:
-bootloader: @oyvindronningstad @hakonfam 
- debug: @nordic-krch 
- esb: @Raane 
- event_manager: @MarekPieta 
- mpsl: @joerchan 
- profiler: @MarekPieta 
- sensor/bh1749: @wlgrd 
- usb/usb_uart_bridge: @bjda 

Should we also align nrf9160?

**Comment by lemrey** [comment]
I think both assertions and stack analyzer could be enabled by default, but we should keep the logger disabled. We are consistently _not_ using the logger to print to the console in the samples to have a consistent behavior (log levels/log buffer sizes/deferred vs immediate) and to hide output from the libs/subsystems onto the console, which I don't think is too interesting in samples in general.

**Comment by rakons** [comment]
If we disable the logger - stack protection (not stack analyzer) and assert would not generate any output. It would just silently go into dead loop - witch generates me a lot of troubles searching for the problem.

**Comment by oyvindronningstad** [comment]
> * bootloader: @oyvindronningstad @hakonfam 

I tried it now, and it causes the bootloader to not fit for pca10040. Will ASSERTs and stack protection be abled to print with printk instead?

**Comment by lemrey** [comment]
I see @rakons, I misread the stack thing.
`__ASSERT` is using `printk`, but from what I could see looking at `zephyr/arch/arm/core/fatal.c` it seems that the fault message is printed with `LOG_ERR`.

**Comment by trond-snekvik** [comment]
This now triggers the massive assert warning in the cmake step, which looks kind of jarring. Any plans to address this?

File: samples/bluetooth/central_bas/prj.conf [hunk 1]
```diff
@@ -3,6 +3,11 @@
 #
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
+CONFIG_ASSERT=y
+CONFIG_ASSERT_LEVEL=2
```
Comment (trond-snekvik) [inline samples/bluetooth/central_bas/prj.conf]: Assert level is 2 by default, skip this entry?

**Comment by joerchan** [comment]
> This now triggers the massive assert warning in the cmake step, which looks kind of jarring. Any plans to address this?

@SebastianBoe @trond-snekvik I don't really know what to do about this warning. It seems a bit excessive to have it. As ASSERTS are very useful and I never understood why compiling them in was such a big deal. It is actually recommended by many to leave asserts in the final release. Is this perhaps something we could address in upstream. AFAIK this is coming from the build system.

**Comment by nordic-krch** [comment]
what about creating default option `APP_LOGGING_DEFAULTS` similar to: https://github.com/zephyrproject-rtos/zephyr/blob/f570f19bef4f18bcedae41f1555a5834e946ce91/subsys/testsuite/Kconfig#L79

Regarding default log options, it can add:
`CONFIG_LOG_MINIMAL=y #logs are converted to printks`
`CONFIG_LOG_MAX_LEVEL=1 #only errors are compiled in`

This should have little impact on memory footprint and none on performance (if asserts are disable).

**Comment by SebastianBoe** [comment]
> I don't really know what to do about this warning. It seems a bit excessive to have it. 

I support dropping the warning, or if we need to compromise, at least make it a one-liner.

**Comment by rakons** [comment]
@joerchan 
> It is actually recommended by many to leave asserts in the final release.

Agree - but you have to configure it properly - to reset the machine, not to dead loop - this way it is better not to compile them in.

@nordic-krch 
> what about creating default option `APP_LOGGING_DEFAULTS`

I like this idea.

@carlescufi 
> That would give the impression that the samples are bloated and slow to the end user?

Why? How much more cycles you need to process assertion checks? In our simple examples it should be hard to notice the difference.

**Comment by SebastianBoe** [comment]
> Why? How much more cycles you need to process assertion checks? In our simple examples it should be hard to notice the difference.

I would be more concerned about the size of the strings in the assert messages, but this is just speculation.

**Comment by joerchan** [comment]
@rakons All valid points. So I think we should leave the asserts in for the samples. It helps us find issues in the samples early. Will you then update to the suggestion by nordic-krch?

@SebastianBoe I think we can afford strings in the samples. They are samples after all. Will you make the change to the ASSERT warning making it oneline instead of five?

**Comment by rakons** [comment]
@joerchan 
Yes - I will update it, but not sure if I manage to do it today. If not - I will handle it at Monday.

**Comment by SebastianBoe** [comment]
> @SebastianBoe I think we can afford strings in the samples. They are samples after all. Will you make the change to the ASSERT warning making it oneline instead of five?

Will do.

BTW, "hello world" flash usage with and without CONFIG_ASSERT:

```
11 216 B
17 040 B
```

more significant than I would have thought, of course, the cost is probably less visible for larger,
more realistic applications.

File: samples/bluetooth/central_bas/prj.conf (hunk 1 above)
Comment (joerchan) [reply samples/bluetooth/central_bas/prj.conf]: Agree, remove if this is the default.

**Comment by rakons** [comment]
Guys, this is how BT_ASSERT is defined: 
```
#define BT_ASSERT(cond) if (!(cond)) { \
				BT_ASSERT_PRINT("assert: '" #cond \
						"' failed\n"); \
				BT_ASSERT_DIE(); \
			}
```
I have to say it is ugly idea. In 99% of usecases I would really prefer to get only a file name and line number. Getting just the condition string gives me no real value. Can we update it?

**Comment by joerchan** [comment]
@rakons That is a **really** good point! Thank you for that. It is actually really bad since it gives us more used code space without providing uniqueness in the assert information. 

I will create a PR to address this issue.

**Comment by lemrey** [comment]
On second thought, more often than not our customers pick the samples and run them as-is to measure different parameters such as power consumption and flash usage. It is a good idea to let samples do what they are supposed to in the most efficient way because that's what the customers first see. The customers will not tweak anything before running a sample because they are / might not be as familiar with the code as we are and they rely on us to provide sane defaults not debugging defaults.

So I changed my mind, and I do not think we should have asserts (or logger) enabled. If we need to debug something, then we can turn on asserts locally.

**Comment by rakons** [comment]
@lemrey 
> parameters such as power consumption

Lets implement some kind of power management first, maybe?

>  flash usage

Then you are saying that it is more important to show the users that our code is small than show them that when they make some mistake and it is not working.

**Comment by lemrey** [comment]
> Then you are saying that it is more important to show the users that our code is small than show them that when they make some mistake and it is not working.

If there is a runtime error in the sample, of course we'll print it; but asserts catch something else.
I am not sure what do you mean by "make a mistake" in a sample, the user is just going to flash it and run it. If a sample is modified by the user, they can enable asserts/logs if something is not working and want to debug.

**Comment by joerchan** [comment]
> > Then you are saying that it is more important to show the users that our code is small than show them that when they make some mistake and it is not working.
> 
> If there is a runtime error in the sample, of course we'll print it; but asserts catch something else.
> I am not sure what do you mean by "make a mistake" in a sample, the user is just going to flash it and run it. If a sample is modified by the user, they can enable asserts/logs if something is not working and want to debug.

@lemrey @rakons I have made memory optimizations to assert handling, this saves a lot in the way assert messages are handled if we enable the extra kconfig option I added.
https://github.com/zephyrproject-rtos/zephyr/pull/21246

**Comment by rakons** [comment]
@joerchan 

I cannot agree that samples are used only to look at it and try to run them. In many situations samples are used as a code start point. Depending on the sample quality - it may be more or less used as a code base.

If you really wish to disable assertions by default - we have to test it anyway on the servers. It is a shame if the user enables assertions and the assertion fails quickly after the system starts.
It is even worst if it fails silently and just goes into dead loop.

**Comment by oyvindronningstad** [comment]
Not arguing either way, but an idea (we do this in the bootloader, b0). The default is debug mode, with printing and everything, but in the prj.conf we have commented out a set of configs with instructions that they will remove debug printing etc. This works the other way as well, so that if logging and asserts are not enabled by default, we can have commented out configs for enabling them.

Longer term, we should have some greater sense of DEBUG and RELEASE builds, if only as a guide in documentation.

**Comment by joerchan** [comment]
To all involved. This discussion will continue offline. Please message me if you have any opinions.

**Comment by grochu** [comment]
I also think assertions should be enabled by default, especially that the NCS samples not only rely on NCS code but also on Zephyr and its dependencies. Upmerges several times introduced problems which sometimes might only be caught using asserts.
Regarding fault handler which ends silently in dead loop (if logging is disabled) - this is a topic for another PR, but perhaps NCS should introduce its own fault handler overriding the Zephyr weak definition. The new handler might reset the chip by default (e.g. when logging is disabled) instead of entering the dead loop.

**Comment by rakons** [comment]
Please recheck - default samples configuration done now in separate file - it should be easier to maintain now.

File: samples/bluetooth/peripheral_hids_keyboard/prj.conf [hunk 2]
```diff
@@ -3,6 +3,8 @@
 #
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
+CONFIG_NCS_SAMPLES_DEFAULTS=y
```
Comment (lemrey) [inline samples/bluetooth/peripheral_hids_keyboard/prj.conf]:
```
CONFIG_ASSERT=y
CONFIG_ASSERT_LEVEL=2
```
Are redundant in this file (see below).

File: samples/bluetooth/peripheral_hids_mouse/prj.conf [hunk 3]
```diff
@@ -3,6 +3,8 @@
 #
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
 #
+CONFIG_NCS_SAMPLES_DEFAULTS=y
```
Comment (lemrey) [inline samples/bluetooth/peripheral_hids_mouse/prj.conf]:
```
CONFIG_ASSERT=y
CONFIG_ASSERT_LEVEL=2
```
Are redundant in this file (see below).

**Comment by rakons** [comment]
@lemrey 
> I think this change should affect all projects under `/samples`, not just `/samples/bluetooth`

I do not know how if the default settings does not broke some of the samples (you can see the comment [here](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1594#issuecomment-559784107) that we cannot use it for bootloader.

I have already added people by nick name that I wished to check theirs examples.

File: samples/bluetooth/peripheral_hids_keyboard/prj.conf (hunk 2 above)
Comment (rakons) [reply samples/bluetooth/peripheral_hids_keyboard/prj.conf]: fixed

File: samples/bluetooth/peripheral_hids_mouse/prj.conf (hunk 3 above)
Comment (rakons) [reply samples/bluetooth/peripheral_hids_mouse/prj.conf]: fixed

**Comment by joerchan** [comment]
> The idea behind this commit is to align the output from the example. To limit the surprise to the user that one example prints something (log enabled) and the other can silently go to dead loop (assertion enabled, log disabled) by default.
> 
> Please look into yours examples and see if you may align the configuration:
> 
> * bootloader: @oyvindronningstad @hakonfam
> * debug: @nordic-krch
> * esb: @Raane
> * event_manager: @MarekPieta
> * mpsl: @joerchan
> * profiler: @MarekPieta
> * sensor/bh1749: @wlgrd
> * usb/usb_uart_bridge: @bjda
> 
> Should we also align nrf9160?

**Comment by joerchan** [comment]
NCSDK-3919 created to follow up for the other samples.

File: samples/Kconfig [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#
+# Copyright (c) 2018 Nordic Semiconductor
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+config NCS_SAMPLES_DEFAULTS
```
Comment (oyvindronningstad) [inline samples/Kconfig]: How will customers relate to this config? What happens if they modify it in menuconfig?  Are they meant to do that? Maybe we should take steps to discourage that if it is not desired? E.g. remove prompt, put a warning in the prompt/help text etc.

Comment (rakons) [reply samples/Kconfig]: Nothing should happen. It would work only when selected during project generation.

Comment (oyvindronningstad) [reply samples/Kconfig]: Right, that makes sense. It should probably be explained in the help text, so they don't think that it is a DEBUG toggle switch or something.

**Pull request merged**